README.md
# Holodeck B2B Signal Notification
This project is an example of a custom extension to the Holodeck B2B Core platform. It implements a _message delivery
method_ for ebMS Signal message units.
The ebMS data from the Signal messages is written to an XML document using the same structure as the ebMS header. A
simple XML schema definition is provided for "cleaner" generated code for XML processing.
As this delivery method is only able to process Signal message units it should only be used in Leg configurations that
 are used for sending User Messages.

__________________
For more information on using Holodeck B2B visit the website at http://holodeck-b2b.org  
Developer: Sander Fieten  
Code hosted at https://github.com/holodeck-b2b/signal-delivery  
Issue tracker https://github.com/holodeck-b2b/signal-delivery/issues

## Installation
Copy the jar file into the `lib` directory of the Holodeck B2B installation **before** the server is started.

### Prerequisites
This extension can be installed in Holodeck B2B version 6.x. As it uses JAXB for XML processing you will need to
install a JAXB implementation when running on Java 9 or later. 

### Configuration
Delivery methods are configured in the P-Mode.
To enable this signal delivery method set the `DeliveryMethod` element to `org.holodeckb2b.backend.file.NotifyOperation`.
The delivery method has two parameters:

1. _targetDirectory_ : should contain the path where the files containing the signal meta-data should be written to;
2. _includeReceiptContent_ : indicates whether the complete content of a _Receipt_ signal should be included in the output
or only the first child element. This is an optional parameter with default value _false_.  

#### Migration
For backward compatibility the old class `org.holodeckb2b.delivery.signals.file.SignalNotifier` and _deliveryDirectory_ 
parameter can still be used. However it is recommended to change your P-Modes to the new class and parameter names as
support for the old names may be removed in the next version.

## Contributing
We’re using the simplified Github workflow to accept modifications which means you’ll have to:
* create an issue related to the problem you want to fix or the function you want to add (good for traceability and cross-reference)
* fork the repository
* create a branch (optionally with the reference to the issue in the name)
* write your code
* commit incrementally with readable and detailed commit messages
* submit a pull-request against the master branch of this repository

### Submitting bugs
Please note that this project is provided as an example and is **not actively supported** by the Holodeck B2B dev team.
You can still report issues on the [project Issue Tracker](https://github.com/holodeck-b2b/signal-delivery/issues)
Please document the steps to reproduce your problem as much as possible.

## Versioning
Version numbering follows the [Semantic versioning](http://semver.org/) approach.

## Licensing
The licence used for this extension is the [European Union Public License](https://joinup.ec.europa.eu/community/eupl/home).
Although Holodeck B2B Core is licensed under the General Public License V3 this is possible because this extension only
uses the *interface module* of the main Holodeck B2B project which is licensed under the Lesser General Public License V3.



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## 2.0.0
##### 2023-03-08
### Changed
* Compatibility with the new Holodeck B2B v6.x API
* Name of the delivery method changed to `org.holodeckb2b.backend.file.NotifyOperation`. 
* Name of the parameter indicating directory where to save Signal meta-data to _targetDirectory_. 

### Deprecated
* The old delivery method name `org.holodeckb2b.delivery.signals.file.SignalNotifier` and its
  _deliveryDirectory_ parameter. 

## 1.1.1
##### 2021-11-22
### Added
* JAXB dependencies to enable building and running on Java 9 or later.

## 1.1.0
##### 2020-09-03
### Added
* Generic method to `org.holodeckb2b.delivery.signals.utils.SMDFactory` to create XML representation of a Signal

### Changed
* Updated dependencies to Holodeck B2B 5.0.0
* When the the _"Include Receipt content"_ parameter is set to _false_ only the first child element of the original 
  Receipt is included without its children.  
 
## 1.0.0
##### 2016-04-08 
### Added
* Initial release. 



src/main/java/org/holodeckb2b/delivery/signals/utils/SMDFactory.java
/* 
* Copyright 2015 The Holodeck B2B Team, Sander Fieten
*  
* Licensed under the EUPL, Version 1.1 or – as soon they will be approved by the European Commission - subsequent
* versions of the EUPL (the "Licence");
* You may not use this work except in compliance with the Licence.
* You may obtain a copy of the Licence at: https://joinup.ec.europa.eu/software/page/eupl
*  
* Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed on 
* an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*
* See the Licence for the specific language governing permissions and limitations under the Licence.
*/ 
package org.holodeckb2b.delivery.signals.utils;

import java.util.Date;
import java.util.GregorianCalendar;
import java.util.Iterator;
import java.util.List;

import javax.xml.datatype.DatatypeConfigurationException;
import javax.xml.datatype.DatatypeFactory;
import javax.xml.datatype.XMLGregorianCalendar;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

import org.apache.axiom.om.OMAttribute;
import org.apache.axiom.om.OMElement;
import org.apache.axiom.om.OMNode;
import org.apache.axiom.om.OMText;
import org.holodeckb2b.delivery.signals.smd.Description;
import org.holodeckb2b.delivery.signals.smd.Error;
import org.holodeckb2b.delivery.signals.smd.MessageInfo;
import org.holodeckb2b.delivery.signals.smd.Receipt;
import org.holodeckb2b.delivery.signals.smd.SignalMessage;
import org.holodeckb2b.interfaces.general.IDescription;
import org.holodeckb2b.interfaces.messagemodel.IEbmsError;
import org.holodeckb2b.interfaces.messagemodel.IErrorMessage;
import org.holodeckb2b.interfaces.messagemodel.IReceipt;
import org.holodeckb2b.interfaces.messagemodel.ISignalMessage;
import org.w3c.dom.Document;
import org.w3c.dom.Element;

/**
 * Is a factory class for creating {@link SignalMessage} objects based on the Holodeck B2B signal message objects.
 * 
 * @author Sander Fieten <sander at holodeck-b2b.org>
 */
public class SMDFactory {
    
    /**
     * Creates a new {@link SignalMessage} object based on the given Holodeck B2B Signal message. When the signal is a
     * Receipt only its content's first child element is included in the XML representation.
     * 
     * @param error    The meta-data about the signal message unit
     * @return         The new {@link SignalMessage} containing the signal meta-data for the given signal
     * @since 1.1.0
     */	
	public static SignalMessage createSMD(final ISignalMessage signal) {
		if (signal instanceof IErrorMessage)
			return createSMD((IErrorMessage) signal);
		else
			return createSMD((IReceipt) signal);
	}
	
    /**
     * Creates a new {@link SignalMessage} object based on the given Holodeck B2B Error signal message.
     * 
     * @param error    The meta-data about the signal message unit as a {@link IErrorMessage}
     * @return         The new {@link SignalMessage} containing the signal meta-data for the given error signal
     */
    public static SignalMessage createSMD(final IErrorMessage error) {
        SignalMessage smd = new SignalMessage();
        
        // Copy the MessageInfo data
        MessageInfo msgInfo = new MessageInfo();
        msgInfo.setMessageId(error.getMessageId());
        msgInfo.setRefToMessageId(error.getRefToMessageId());        
        msgInfo.setTimestamp(convertDate(error.getTimestamp()));        
        smd.setMessageInfo(msgInfo);
        
        // Add the error information to the SMD
        List<Error> errors = smd.getError();
        for (IEbmsError e : error.getErrors()) 
            errors.add(createErrorElement(e));
        
        return smd;
    }
    
    /**
     * Helper method to create an Error element in the SMD based on the Error element from the ebMS header.
     * 
     * @param error     The error from the ebMS header
     * @return          The error meta-data to include in the SMD
     */
    protected static Error createErrorElement(final IEbmsError error) {
        Error   err = new Error();
        
        err.setErrorCode(error.getErrorCode());
        err.setSeverity(error.getSeverity().name());
        err.setErrorDetail(error.getErrorDetail());
        err.setShortDescription(error.getMessage());
        err.setCategory(error.getCategory());
        err.setOrigin(error.getOrigin());
        
        // Copy the extended description when available
        IDescription eDescr = error.getDescription();
        if (eDescr != null && eDescr.getText() != null && !eDescr.getText().isEmpty()) {
            Description description = new Description();
            description.setValue(eDescr.getText());
            description.setLang(eDescr.getLanguage());
            
            err.setDescription(description);
        }
            
        return err;
    }
    
    /**
     * Creates a new {@link SignalMessage} object based on the given Holodeck B2B Receipt signal message with only the
     * first child of the original Receipt included in the notification so the business application can determine what
     * kind of Receipt was received.
     * 
     * @param receipt  The meta-data about the signal message unit as a {@link IReceipt}
     * @return         The new {@link SignalMessage} containing the signal meta-data for the given receipt signal
     */
    public static SignalMessage createSMD(final IReceipt receipt) {
        return createSMD(receipt, false);
    }
    
    
    /**
     * Creates a new {@link SignalMessage} object based on the given Holodeck B2B Receipt signal message.
     * 
     * @param receipt               The meta-data about the signal message unit as a {@link IReceipt}
     * @param includeFullContent    Indicates whether the complete content from the Receipt as in the ebMS message
     *                              should be included in the SMD.
     * @return         The new {@link SignalMessage} containing the signal meta-data for the given receipt signal
     */
    public static SignalMessage createSMD(final IReceipt receipt, final boolean includeFullContent) {
        SignalMessage smd = new SignalMessage();
        
        // Copy the MessageInfo data
        MessageInfo msgInfo = new MessageInfo();
        msgInfo.setMessageId(receipt.getMessageId());
        msgInfo.setRefToMessageId(receipt.getRefToMessageId());        
        msgInfo.setTimestamp(convertDate(receipt.getTimestamp()));        
        smd.setMessageInfo(msgInfo);
        
        // Now copy the Receipt data
        Receipt         rcpt = new Receipt();
        List<Object>    content = rcpt.getAny();        
        List<OMElement> rcptContent = receipt.getContent();
        
        DocumentBuilder builder;
        try {
            builder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
        } catch (ParserConfigurationException ex) {
            // We can not create a valid SMD if we can not create an element to contain in the Receipt element
            return null;
        }
        Document doc = builder.newDocument();
        
        if (includeFullContent) {
            // Copy all child elements of the Receipt
            for (OMElement e : rcptContent)
                content.add(convertElement(e, doc, true));
        } else {
            // Only the first child element of the Receipt needs to be included
            content.add(convertElement(rcptContent.get(0), doc, false));            
        }
        
        // Add Receipt info to SignalMessage
        smd.setReceipt(rcpt);
            
        return smd;
    }
    
    
    /**
     * Helper method to convert a {@link Date} object to a {@link XMLGregorianCalendar} object.
     * 
     * @param date  The date to convert as {@link Date}
     * @return      The data as {@link XMLGregorianCalendar} if it could be converted, <code>null</code> otherwise
     */
    private static XMLGregorianCalendar convertDate(final Date date) {
        try {
            GregorianCalendar c = new GregorianCalendar();
            c.setTime(date);
            return DatatypeFactory.newInstance().newXMLGregorianCalendar(c);
        } catch (DatatypeConfigurationException ex) {
            return null;
        }
    }
    
    /**
     * Helper method to convert an Axiom {@link OMElement} to a W3 DOM representation.
     * 
     * @param e     	The Axiom representation of the element
     * @param doc   	The W3 Document in which the DOM representation is to be inserted
     * @param deepCopy 	Indicates whether the child elements should be included in the converted element
     * @return      The converted element
     */
    private static Element convertElement(OMElement e, Document doc, boolean deepCopy) {
        // Convert the element itself
        Element d = doc.createElementNS(e.getNamespaceURI(), e.getLocalName());
        String prefix = e.getPrefix();
        if (prefix != null)
            d.setPrefix(prefix);
        // And all its attributes
        Iterator<OMAttribute> attrs = e.getAllAttributes();
        for(OMAttribute a ; attrs.hasNext() ;) {
            a = attrs.next();
            String ns = a.getNamespaceURI();
            if (ns != null)
                d.setAttributeNS(ns, a.getLocalName(), a.getAttributeValue());
            else
                d.setAttribute(a.getLocalName(), a.getAttributeValue());
        }
        if (deepCopy) {
	        // And convert all content child nodes, calling this method recursively for child elements
	        Iterator<OMNode> content = e.getChildren();
	        while (content.hasNext()) {
	            OMNode n = content.next();
	            if (n instanceof OMElement)
	                d.appendChild(convertElement((OMElement) n, doc, true));
	            else if (n instanceof OMText)
	                d.appendChild(doc.createTextNode(((OMText) n).getText()));
	        }
        }
            
        return d;
    }
}


src/main/java/org/holodeckb2b/delivery/signals/file/SignalNotifier.java
/* 
* Copyright 2015 The Holodeck B2B Team, Sander Fieten
*  
* Licensed under the EUPL, Version 1.1 or – as soon they will be approved by the European Commission - subsequent
* versions of the EUPL (the "Licence");
* You may not use this work except in compliance with the Licence.
* You may obtain a copy of the Licence at: https://joinup.ec.europa.eu/software/page/eupl
*  
* Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed on 
* an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*
* See the Licence for the specific language governing permissions and limitations under the Licence.
*/ 
package org.holodeckb2b.delivery.signals.file;

import java.util.AbstractMap;
import java.util.Map;
import java.util.stream.Collectors;

import org.apache.logging.log4j.LogManager;
import org.holodeckb2b.backend.file.NotifyOperation;
import org.holodeckb2b.interfaces.delivery.MessageDeliveryException;

/**
 * @deprecated This class is only included for back-ward compatibility. Configurations <b>should</b> be changed asap.
 */
@Deprecated
public class SignalNotifier extends NotifyOperation {

	@Override
	public void init(Map<String, ?> settings) throws MessageDeliveryException {
		LogManager.getLogger().warn("org.holodeckb2b.delivery.signals.SignalNotifier is deprecated! Use org.holodeckb2b.backend.file.NotifyOperation instead");		
		super.init(settings.entrySet().stream().map(e -> "deliveryDirectory".equals(e.getKey()) ?
													new AbstractMap.SimpleEntry<>("targetDirectory", e.getValue()) : e)
							.collect(Collectors.toMap(r -> r.getKey(), r -> r.getValue())));
	}	   
}


src/main/java/org/holodeckb2b/delivery/signals/file/SignalNotifierDeliverer.java
/*
* Copyright 2015 The Holodeck B2B Team, Sander Fieten
*
* Licensed under the EUPL, Version 1.1 or – as soon they will be approved by the European Commission - subsequent
* versions of the EUPL (the "Licence");
* You may not use this work except in compliance with the Licence.
* You may obtain a copy of the Licence at: https://joinup.ec.europa.eu/software/page/eupl
*
* Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed on
* an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*
* See the Licence for the specific language governing permissions and limitations under the Licence.
*/
package org.holodeckb2b.delivery.signals.file;

import java.io.File;
import java.io.IOException;
import java.nio.file.FileAlreadyExistsException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import javax.xml.bind.JAXBContext;
import javax.xml.bind.JAXBElement;
import javax.xml.bind.JAXBException;
import javax.xml.bind.Marshaller;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.holodeckb2b.delivery.signals.smd.ObjectFactory;
import org.holodeckb2b.delivery.signals.smd.SignalMessage;
import org.holodeckb2b.delivery.signals.utils.SMDFactory;
import org.holodeckb2b.interfaces.delivery.IMessageDeliverer;
import org.holodeckb2b.interfaces.delivery.MessageDeliveryException;
import org.holodeckb2b.interfaces.messagemodel.IErrorMessage;
import org.holodeckb2b.interfaces.messagemodel.IMessageUnit;
import org.holodeckb2b.interfaces.messagemodel.IReceipt;

/**
 * Is the {@link IMessageDeliverer} implementation for notifying business applications about signal messages.
 *
 * @author Sander Fieten <sander at holodeck-b2b.org>
 */
public class SignalNotifierDeliverer implements IMessageDeliverer {

    private final Logger log = LogManager.getLogger(SignalNotifierDeliverer.class);

    /**
     * The directory where to write the signal meta-data file
     */
    private String  deliveryDir;

    /**
     * Indicates whether the content of the Receipt should be included in the notification. Default it is not.
     */
    private boolean includeReceiptContent;

    /**
     * Creates the deliver and sets the destination directory for the meta-data documents and indicator whether the
     * Receipt content should be included in the notifications.
     *
     * @param deliveryDir               Path of the directory where to write the signal meta-data documents.
     * @param includeReceiptContent     Indication whether to include the Receipt content in the notification.
     */
    public SignalNotifierDeliverer(final String deliveryDir, final boolean includeReceiptContent) {
        this.deliveryDir = deliveryDir;
        this.includeReceiptContent = includeReceiptContent;
    }

    /**
     * Does the actual delivery of the signal message to the business application.
     *
     * @param rcvdMsgUnit   The signal message to be delivered to the business application
     * @throws MessageDeliveryException When the given message unit is not a signal or when the signal meta-data
     *                                  document can not be created or written to file.
     */
    @Override
    public void deliver(IMessageUnit rcvdMsgUnit) throws MessageDeliveryException {
        String sigType = rcvdMsgUnit.getClass().getSimpleName();
        String sigMsgId = rcvdMsgUnit.getMessageId();

        log.debug("Create SMD for " + sigType + ", msgId= " + sigMsgId);
        SignalMessage smd;
        if (rcvdMsgUnit instanceof IReceipt)
            smd = SMDFactory.createSMD((IReceipt) rcvdMsgUnit, includeReceiptContent);
        else if (rcvdMsgUnit instanceof IErrorMessage)
            smd = SMDFactory.createSMD((IErrorMessage) rcvdMsgUnit);
        else {
            log.warn("This delivery method can not be used for delivery of User Messages!");
            throw new MessageDeliveryException("This delivery method can not be used for User messages!");
        }

        // Check that SMD document could be created
        if (smd == null) {
            log.error("Failed to create the SMD for " + sigType + ", msgId= " + sigMsgId);
            throw new MessageDeliveryException("Could not create meta-data document for Signal message");
        }

        // Create the filename for the SMD file based on message id of signal
        File file = null;
        try {
            file = ensureUniqueFile(deliveryDir + sigMsgId.replaceAll("[^a-zA-Z0-9.-]", "_") + ".smd.xml");
        } catch (IOException ex) {
            log.error("Can not create a file in output directory {} to write Signal meta-data (msgId={})",
                        deliveryDir, sigMsgId);
            throw new MessageDeliveryException("Could not create file for delivery of Signal", ex);
        }
        log.debug("Write SMD for {} (msgId= {}) to {}", sigType, sigMsgId, file.getAbsolutePath());

        try {
            JAXBContext jbCtx = JAXBContext.newInstance(SignalMessage.class.getPackage().getName());
            Marshaller jbMarshaller = jbCtx.createMarshaller();
            JAXBElement<SignalMessage> smdDoc = new ObjectFactory().createSignalMessage(smd);
            jbMarshaller.setProperty(Marshaller.JAXB_FORMATTED_OUTPUT, true);
            jbMarshaller.marshal(smdDoc, file);
        } catch (JAXBException e) {
            log.error("Could not write SMD for " + sigType + "(msgId= " + sigMsgId + ")! Details: "
                        + e.getErrorCode() + " - " + e.getMessage());
            throw new MessageDeliveryException("Could not write the SMD file!", e);
        }
    }

    /**
     * Helper method to ensure the file for writing the Signal to is unique.
     *
     * @param basePath      The default file name to use for writing the Signal meta-data
     * @return              A {@link File} object for the file that the meta-data can be written to
     * @throws IOException  When the file can not be created
     */
    private File ensureUniqueFile(final String basePath) throws IOException {
        // Split the given path into name and extension part (if possible)
        String nameOnly = basePath;
        String ext = "";
        int startExt = basePath.lastIndexOf(".");
        if (startExt > 0) {
            nameOnly = basePath.substring(0, startExt);
            ext = basePath.substring(startExt);
        }

        File f = null;
        Path targetPath = Paths.get(basePath);
        int i = 1;
        while (f == null) {
            try {
                f = Files.createFile(targetPath).toFile();
            } catch (FileAlreadyExistsException faee) {
                // Okay, the file already exists, try with increased sequence number
                targetPath = Paths.get(nameOnly + "-" + i++ + ext);
            }
        }

        return f;
    }

}


src/main/java/org/holodeckb2b/backend/file/NotifyOperation.java
/* 
* Copyright 2023 The Holodeck B2B Team, Sander Fieten
*  
* Licensed under the EUPL, Version 1.1 or – as soon they will be approved by the European Commission - subsequent
* versions of the EUPL (the "Licence");
* You may not use this work except in compliance with the Licence.
* You may obtain a copy of the Licence at: https://joinup.ec.europa.eu/software/page/eupl
*  
* Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed on 
* an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*
* See the Licence for the specific language governing permissions and limitations under the Licence.
*/ 
package org.holodeckb2b.backend.file;

import java.io.IOException;
import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Map;

import javax.xml.bind.JAXBContext;
import javax.xml.bind.JAXBElement;
import javax.xml.bind.JAXBException;
import javax.xml.bind.Marshaller;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.holodeckb2b.commons.util.FileUtils;
import org.holodeckb2b.commons.util.Utils;
import org.holodeckb2b.delivery.signals.smd.ObjectFactory;
import org.holodeckb2b.delivery.signals.smd.SignalMessage;
import org.holodeckb2b.delivery.signals.utils.SMDFactory;
import org.holodeckb2b.interfaces.delivery.IDeliveryMethod;
import org.holodeckb2b.interfaces.delivery.MessageDeliveryException;
import org.holodeckb2b.interfaces.messagemodel.IErrorMessage;
import org.holodeckb2b.interfaces.messagemodel.IMessageUnit;
import org.holodeckb2b.interfaces.messagemodel.IReceipt;

/**
 * Is a {@link IDeliveryMethod} implementation that notifies the business application about received <i>Signal 
 * Messages</i> by writing the signal meta-data (SMD) to file.
 * <p>The XML format of the meta data document is based on the ebMS header for signals and is defined in the 
 * <code>http://holodeck-b2b.org/schemas/2015/08/smd</code> xml schema definition.
 * <p>The delivery method takes two parameters:<ol>
 * <li><i>targetDirectory</i> : the path to the directory where the SMD files should be written.</li>
 * <li><i>includeReceiptContent</i> : a boolean that indicate whether the complete Receipt content from the ebMS message
 * should be included (<i>"true"</i>) or only the first child element (<i>"false"</i>). Default is <i>false</i> (only
 * first child included).</li>
 * </ol>
 * <p>NOTE: This delivery method can only be used for Signals!
 */
public class NotifyOperation implements IDeliveryMethod {
    private final Logger log = LogManager.getLogger(NotifyOperation.class);

    /**
     * The path to the directory where to save the notifications
     */
    private String  deliveryDir = null;
    
    /**
     * Indicates whether the content of the Receipt should be included in the notification. Default it is not.
     */
    private boolean includeReceiptContent = false;
	
    /**
     * Initializes the delivery method. Ensures that the specified directory is available for delivery of the Signals,
     * i.e. if the path already exists checks that it is a writable directory or if it does not exist yet create it.
     * 
     * @param settings  The settings to use for the notifications. MUST contain at least on entry with key 
     *                  <code>deliveryDirectoy</code> holding the path to the delivery directory. 
     * @throws MessageDeliveryException When there the specified directory is not writable or can not be created
     */
	@Override
	public void init(Map<String, ?> settings) throws MessageDeliveryException {
        String includeRcptCont = null;
        
        if (settings != null) {
            try {
               deliveryDir = (String) settings.get("targetDirectory");                
               includeRcptCont = (String) settings.get("includeReceiptContent");
            } catch (ClassCastException ex) {
                throw new MessageDeliveryException("Configuration error! No directory specified!");
            }
        }        
        if (!checkDirectory())
            throw new MessageDeliveryException("Configuration error! Specified directory [" + deliveryDir
                                                                        + "] is not available!");
        // Ensure directory path ends with separator
        deliveryDir = (deliveryDir.endsWith(FileSystems.getDefault().getSeparator()) ? deliveryDir 
                              : deliveryDir + FileSystems.getDefault().getSeparator());
        
        // Should we include receipt content?
        includeReceiptContent = Utils.isTrue(includeRcptCont);
	}

	 /**
     * Helper method that checks if the specified path for delivery of the Signals is a writable directory or if the
     * path does not exist a directory can be created.
     * 
     * @return  <code>true</code> when the specified path can be used for delivery,<br><code>false</code> otherwise
     */
    private boolean checkDirectory() {
        try {
            Path path = Paths.get(deliveryDir);                        

            // Test if given path exists and is a writable directory
            if (Files.exists(path))
                // The path exists, it can be used for delivery if it is a writable directory
                return Files.isDirectory(path) && Files.isWritable(path);
            else {
                // The path does not exist yet, so try to create it
                Files.createDirectories(path);            
                return true;
            }
        } catch (Exception ex) {
            return false;
        }
    }
    
	@Override
	public boolean supportsAsyncDelivery() {
		return false;
	}

	/**
     * Does the actual delivery of the signal message to the business application.
     *
     * @param rcvdMsgUnit   The signal message to be delivered to the business application
     * @throws MessageDeliveryException When the given message unit is not a signal or when the signal meta-data
     *                                  document can not be created or written to file.
     */
    @Override
    public void deliver(IMessageUnit rcvdMsgUnit) throws MessageDeliveryException {
        String sigType = rcvdMsgUnit.getClass().getSimpleName();
        String sigMsgId = rcvdMsgUnit.getMessageId();

        log.debug("Create SMD for " + sigType + ", msgId= " + sigMsgId);
        SignalMessage smd;
        if (rcvdMsgUnit instanceof IReceipt)
            smd = SMDFactory.createSMD((IReceipt) rcvdMsgUnit, includeReceiptContent);
        else if (rcvdMsgUnit instanceof IErrorMessage)
            smd = SMDFactory.createSMD((IErrorMessage) rcvdMsgUnit);
        else {
            log.warn("This delivery method can not be used for delivery of User Messages!");
            throw new MessageDeliveryException("This delivery method can not be used for User messages!");
        }

        // Check that SMD document could be created
        if (smd == null) {
            log.error("Failed to create the SMD for " + sigType + ", msgId= " + sigMsgId);
            throw new MessageDeliveryException("Could not create meta-data document for Signal message");
        }

        // Create the filename for the SMD file based on message id of signal
        Path file = null;
        try {
            file = FileUtils.createFileWithUniqueName(deliveryDir + sigMsgId.replaceAll("[^a-zA-Z0-9.-]", "_") + ".smd.xml");
        } catch (IOException ex) {
            log.error("Can not create a file in output directory {} to write Signal meta-data (msgId={})",
                        deliveryDir, sigMsgId);
            throw new MessageDeliveryException("Could not create file for delivery of Signal", ex);
        }
        log.debug("Write SMD for {} (msgId= {}) to {}", sigType, sigMsgId, file.toString());

        try {
            JAXBContext jbCtx = JAXBContext.newInstance(SignalMessage.class.getPackage().getName());
            Marshaller jbMarshaller = jbCtx.createMarshaller();
            JAXBElement<SignalMessage> smdDoc = new ObjectFactory().createSignalMessage(smd);
            jbMarshaller.setProperty(Marshaller.JAXB_FORMATTED_OUTPUT, true);
            jbMarshaller.marshal(smdDoc, file.toFile());
        } catch (JAXBException e) {
            log.error("Could not write SMD for " + sigType + "(msgId= " + sigMsgId + ")! Details: "
                        + e.getErrorCode() + " - " + e.getMessage());
            throw new MessageDeliveryException("Could not write the SMD file!", e);
        }
    }   
}
